src/Microsoft.Azure.WebJobs.Shared.StorageProvider/StorageClientProvider.cs

```diff
@@ -0,0 +1,230 @@
+        /// <returns>storage client provider</returns>
+        public virtual TClient Get(string name)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                name = ConnectionStringNames.Storage; // default
+            }
+
+            // $$$ Where does validation happen?
+            IConfigurationSection connectionSection = _configuration.GetWebJobsConnectionStringSection(name);
+            if (!connectionSection.Exists())
+            {
+                // Not found
+                throw new InvalidOperationException($"Storage account connection string '{IConfigurationExtensions.GetPrefixedConnectionStringName(name)}' does not exist. Make sure that it is a defined App Setting.");
+            }
+
+            _logForwarder.Start();
+            var credential = _componentFactory.CreateTokenCredential(connectionSection);
+            var options = CreateClientOptions(connectionSection);
+            return CreateClient(connectionSection, credential, options);
+        }
+
+        /// <summary>
+        /// Creates a storage client
+        /// </summary>
+        /// <param name="configuration">Registered <see cref="IConfiguration"/></param>
+        /// <param name="tokenCredential"><see cref="TokenCredential"/> to use for the client </param>
+        /// <param name="options">Generic options to use for the client</param>
+        /// <returns></returns>
+        protected virtual TClient CreateClient(IConfiguration configuration, TokenCredential tokenCredential, TClientOptions options)
+        {
+            // If connection string is present, it will be honored first
+            if (!IsConnectionStringPresent(configuration) && TryGetServiceUri(configuration, out Uri serviceUri))
+            {
+                var constructor = typeof(TClient).GetConstructor(new Type[] { typeof(Uri), typeof(TokenCredential), typeof(TClientOptions) });
+                return (TClient)constructor.Invoke(new object[] { serviceUri, tokenCredential, options });
+            }
+
+            return (TClient)_componentFactory.CreateClient(typeof(TClient), configuration, tokenCredential, options);
+        }
+
+        /// <summary>
+        /// The host account is for internal storage mechanisms like load balancer queuing.
+        /// </summary>
+        /// <returns></returns>
+        public virtual TClient GetHost()
+        {
+            return this.Get(null);
+        }
+
+        /// <summary>
+        /// Creates client options from the given configuration
+        /// </summary>
+        /// <param name="configuration">Registered <see cref="IConfiguration"/></param>
+        /// <returns>Client options</returns>
+        protected virtual TClientOptions CreateClientOptions(IConfiguration configuration)
+        {
+            var clientOptions = (TClientOptions)_componentFactory.CreateClientOptions(typeof(TClientOptions), null, configuration);
```

Comment:
This logic in componentFactory is already shipped and handles all the client creation. It's probably best to unify client creation in the componentFactory rather than creating it ourselves (unless we have a different configuration like the serviceUri specification)

Reply:
I don't think that addresses the initial question. This is for creation of the options class, not the client. Is there a reason this type is not just taking a dependency on `IOptions<TClientOptions>` instead?

Reply:
BlobClientOptions and QueueClientOptions need parameterless constructors to use with the Options pattern. Not possible here.